File: app/src/main/java/de/uni_marburg/mathematik/ds/serval/view/fragments/MapFragment.java
package de.uni_marburg.mathematik.ds.serval.view.fragments;

import android.content.Intent;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMap.OnMyLocationButtonClickListener;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.maps.android.clustering.Cluster;
import com.google.maps.android.clustering.ClusterItem;
import com.google.maps.android.clustering.ClusterManager;
import com.google.maps.android.clustering.ClusterManager.OnClusterClickListener;
import com.google.maps.android.clustering.ClusterManager.OnClusterItemInfoWindowClickListener;

import de.uni_marburg.mathematik.ds.serval.R;
import de.uni_marburg.mathematik.ds.serval.controller.adapters.ExtendedInfoWindowAdapter;
import de.uni_marburg.mathematik.ds.serval.model.event.Event;
import de.uni_marburg.mathematik.ds.serval.model.event.EventCallback;
import de.uni_marburg.mathematik.ds.serval.util.PrefManager;
import de.uni_marburg.mathematik.ds.serval.view.activities.DetailActivity;
import de.uni_marburg.mathematik.ds.serval.view.activities.MainActivity;

import static android.Manifest.permission.ACCESS_FINE_LOCATION;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;
import static android.support.v4.content.ContextCompat.checkSelfPermission;

/**
 * TODO Add JavaDoc
 *
 * @param <T>
 */
public class MapFragment<T extends Event>
        extends EventFragment<T>
        implements OnClusterItemInfoWindowClickListener<T>, OnClusterClickListener<T>,
                   OnMapReadyCallback, OnMyLocationButtonClickListener {
    
    private static final int CHECK_LOCATION_PERMISSION = 42;
    
    private static final int MAP_PADDING = 50;
    
    /**
     * 1: World
     * 5: Continent
     * 10: City
     * 15: Streets
     * 20: Buildings
     */
    private static final float MAP_ZOOM = 15;
    
    private ClusterManager<T> clusterManager;
    
    private GoogleMap googleMap;
    
    private Location lastLocation;
    
    private PrefManager prefManager;
    
    private SupportMapFragment map;
    
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }
    
    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        map = (SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.map);
    }
    
    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setupFields();
        map.getMapAsync(this);
    }
    
    @Override
    protected int getLayout() {
        return R.layout.fragment_map;
    }
    
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_map, menu);
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_change_map_type_hybrid:
                googleMap.setMapType(GoogleMap.MAP_TYPE_HYBRID);
                return true;
            case R.id.action_change_map_type_none:
                googleMap.setMapType(GoogleMap.MAP_TYPE_NONE);
                return true;
            case R.id.action_change_map_type_normal:
                googleMap.setMapType(GoogleMap.MAP_TYPE_NORMAL);
                return true;
            case R.id.action_change_map_type_satellite:
                googleMap.setMapType(GoogleMap.MAP_TYPE_SATELLITE);
                return true;
            case R.id.action_change_map_type_terrain:
                googleMap.setMapType(GoogleMap.MAP_TYPE_TERRAIN);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
    
    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.googleMap = googleMap;
        setupGoogleMap();
        if (prefManager.requestLocationUpdates()) {
            moveCameraToLastLocation(false);
        }
    }
    
    @Override
    public boolean onMyLocationButtonClick() {
        moveCameraToLastLocation(true);
        return true;
    }
    
    @Override
    public boolean onClusterClick(Cluster<T> cluster) {
        LatLngBounds.Builder builder = LatLngBounds.builder();
        for (ClusterItem clusterItem : cluster.getItems()) {
            builder.include(clusterItem.getPosition());
        }
        googleMap.animateCamera(CameraUpdateFactory.newLatLngBounds(builder.build(), MAP_PADDING));
        return true;
    }
    
    @Override
    public void onClusterItemInfoWindowClick(T event) {
        Intent eventIntent = new Intent(getActivity(), DetailActivity.class);
        eventIntent.putExtra(DetailActivity.EVENT, event);
        startActivity(eventIntent);
    }
    
    private void setupFields() {
        //noinspection unchecked
        eventCallback = (EventCallback<T>) getActivity();
        lastLocation = ((MainActivity) getActivity()).getLastLocation();
        prefManager = new PrefManager(getContext());
    }
    
    private void setupGoogleMap() {
        setupClusterManager();
        setupGoogleMapsListeners(clusterManager);
        setupCameraBounds();
        setupMyLocation();
    }
    
    private void setupClusterManager() {
        clusterManager = new ClusterManager<>(getContext(), googleMap);
        clusterManager.setOnClusterClickListener(this);
        clusterManager.setOnClusterItemInfoWindowClickListener(this);
        clusterManager.addItems(events);
        clusterManager.cluster();
    }
    
    private void setupGoogleMapsListeners(ClusterManager<T> clusterManager) {
        googleMap.getUiSettings().setMapToolbarEnabled(false);
        googleMap.setOnCameraIdleListener(clusterManager);
        googleMap.setOnMarkerClickListener(clusterManager);
        googleMap.setOnInfoWindowClickListener(clusterManager);
        googleMap.setInfoWindowAdapter(new ExtendedInfoWindowAdapter(getContext()));
    }
    
    private void setupCameraBounds() {
        LatLngBounds.Builder builder = LatLngBounds.builder();
        for (T event : events) {
            builder.include(event.getPosition());
        }
        googleMap.setLatLngBoundsForCameraTarget(builder.build());
    }
    
    private void setupMyLocation() {
        if (prefManager.requestLocationUpdates()) {
            if (checkSelfPermission(getContext(), ACCESS_FINE_LOCATION) != PERMISSION_GRANTED) {
                requestPermissions();
                return;
            }
            googleMap.setMyLocationEnabled(true);
            googleMap.setOnMyLocationButtonClickListener(this);
        }
    }
    
    private void requestPermissions() {
        ActivityCompat.requestPermissions(
                getActivity(),
                new String[]{ACCESS_FINE_LOCATION},
                CHECK_LOCATION_PERMISSION
        );
    }
    
    private void moveCameraToLastLocation(boolean animate) {
        if (lastLocation != null) {
            LatLng lastLocationPosition = new LatLng(
                    lastLocation.getLatitude(),
                    lastLocation.getLongitude()
            );
            CameraUpdate update = CameraUpdateFactory.newLatLngZoom(lastLocationPosition, MAP_ZOOM);
            if (animate) {
                googleMap.animateCamera(update);
            } else {
                googleMap.moveCamera(update);
            }
        }
    }
}

File: app/src/main/java/de/uni_marburg/mathematik/ds/serval/view/fragments/MapFragment.kt
package de.uni_marburg.mathematik.ds.serval.view.fragments

import android.Manifest
import android.annotation.SuppressLint
import android.graphics.PorterDuff
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import ca.allanwang.kau.utils.color
import ca.allanwang.kau.utils.drawable
import ca.allanwang.kau.utils.hasPermission
import com.google.android.gms.maps.CameraUpdate
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.*
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLngBounds
import com.google.maps.android.clustering.ClusterManager
import de.uni_marburg.mathematik.ds.serval.R
import de.uni_marburg.mathematik.ds.serval.model.event.Event
import de.uni_marburg.mathematik.ds.serval.util.MAP_PADDING
import de.uni_marburg.mathematik.ds.serval.util.consume
import de.uni_marburg.mathematik.ds.serval.view.activities.DetailActivity
import de.uni_marburg.mathematik.ds.serval.view.activities.MainActivity
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.uiThread

class MapFragment : BaseFragment() {

    private lateinit var googleMap: GoogleMap

    private val map: SupportMapFragment by lazy {
        childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment
    }

    override val layout: Int
        get() = R.layout.fragment_map

    @SuppressLint("MissingPermission")
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        map.getMapAsync { googleMap ->
            this.googleMap = googleMap
            with(googleMap) {
                if (activity!!.hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)) {
                    isMyLocationEnabled = true
                }
                uiSettings.isMapToolbarEnabled = false
                setupClusterManager()
                setupCamera()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater?.inflate(R.menu.menu_map, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        menu?.let {
            super.onPrepareOptionsMenu(menu)
            val changeMapType = menu.findItem(R.id.action_change_map_type)
            val icon = context!!.drawable(R.drawable.map)
            icon.setColorFilter(context!!.color(android.R.color.white), PorterDuff.Mode.SRC_IN)
            changeMapType.icon = icon
        }
    }

    override fun onOptionsItemSelected(item: MenuItem) = with(googleMap) {
        when (item.itemId) {
            R.id.action_change_map_type_hybrid -> consume { mapType = MAP_TYPE_HYBRID }
            R.id.action_change_map_type_none -> consume { mapType = MAP_TYPE_NONE }
            R.id.action_change_map_type_normal -> consume { mapType = MAP_TYPE_NORMAL }
            R.id.action_change_map_type_satellite -> consume { mapType = MAP_TYPE_SATELLITE }
            R.id.action_change_map_type_terrain -> consume { mapType = MAP_TYPE_TERRAIN }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun GoogleMap.setupCamera() {
        if (MainActivity.events.isNotEmpty()) {
            val builder = LatLngBounds.builder()
            doAsync {
                MainActivity.events.forEach { builder.include(it.position) }
                uiThread {
                    with(builder.build()) {
                        googleMap.setLatLngBoundsForCameraTarget(this)
                        googleMap.cameraUpdate(this)
                    }
                }
            }
        }
    }

    private fun GoogleMap.setupClusterManager() {
        with(ClusterManager<Event>(context, this)) {
            setAnimation(false)
            setOnCameraIdleListener(this)
            setOnMarkerClickListener(this)
            setOnInfoWindowClickListener(this)
            setOnClusterClickListener { cluster ->
                consume {
                    with(LatLngBounds.builder()) {
                        cluster.items.forEach { event -> include(event.position) }
                        cameraUpdate(build(), true)
                    }
                }
            }
            setOnClusterItemInfoWindowClickListener { event ->
                context!!.startActivity<DetailActivity>(DetailActivity.EVENT to event)
            }

            doAsync {
                with(MainActivity.events) {
                    addItems(this)
                    uiThread { cluster() }
                }
            }
        }
    }

    private fun GoogleMap.cameraUpdate(bounds: LatLngBounds, animate: Boolean = false) {
        val cameraUpdate: CameraUpdate = CameraUpdateFactory.newLatLngBounds(bounds, MAP_PADDING)
        if (animate) animateCamera(cameraUpdate) else moveCamera(cameraUpdate)
    }

}
